Fixed ZoomPanBase to work with log plots

ordirules opened the pull request:
Tested this code in a matplotlib 1.4.3 environment and current branch. Fixed ZoomPanBase to work in non-linear plots by changing the scale in the axes coordinates and using the respective tranformation functions: transScale and transLimits transforms. 

See tutorial on axes transformations for more details:
http://matplotlib.org/users/transforms_tutorial.html

Note: It did not work in my version 1.4.2. The issue is traceable to the TransScale.transform.inverted() routine expecting a 2 dimensional array rather than 1d array. If applying this to this older version this should be investigated/tested.

Testing code used (use mouse scroll wheel afterwards):
```python
import matplotlib.pyplot as plt
import matplotlib.backend_tools as bt
import numpy as np

fig1 = plt.figure(0)
zm = bt.ZoomPanBase(fig1,"foo")


img = np.random.random((100,100))
ax = fig1.add_subplot(111)

x = np.linspace(10,1000,1000)
y = 1/x**4

#uncomment the one you want to test, and comment the other
#ax.imshow(img)
ax.loglog(x,y)

def foofunc():
    pass

zm._press = foofunc
zm._release = foofunc

zm.enable("foo")
```

tacaswell commented:
attn @fariza @OceanWolf Can you two take care of reviewing and merging this?

ordirules commented:
Thomas, can we hold off for a bit? I just realized that my zooming does not conform to yours.
(Zoom center is center of axes, not the cursor)

Will submit the correction and test when I have time. Thanks

ordirules commented:
@tacaswell 
There are three possible zooming conventions:
1. Zoom center is always center of figure (my current version)
2. Zoom center is cursor. Once zoomed, the point under cursor is moved to center of image (current version in this library)
3. Zoom center is cursor. Upon zooming, the point under the cursor does not move.

Which one to use? I prefer number 3. Have code ready, will submit change once convention is agreed upon.

WeatherGod commented:
I have always disliked how zooming happened in matplotlib. I vote for
option 3.

On Thu, Aug 20, 2015 at 10:50 AM, Julien Lhermitte <notifications@github.com
> wrote:

> @tacaswell <https://github.com/tacaswell>
> There are three possible zooming conventions:
> 1. Zoom center is always center of figure (my current version)
> 2. Zoom center is cursor. Once zoomed, the point under cursor is moved to
> center of image (current version in this library)
> 3. Zoom center is cursor. Upon zooming, the point under the cursor does
> not move.
>
> Which one to use? I prefer number 3. Have code ready, will submit change
> once convention is agreed upon.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4970#issuecomment-133038777>
> .
>

ordirules commented:
Two votes is good for me. Submitted. Tested on this branch with same code mentioned above.

(math : new axis position: (xa -/+.5*scl + (.5-xa)*scl, then use matplotlib functions to convert to data, which take care of non-linear plots.)

fariza commented:
I'll check this next week. I'm on vacation :)
On 20 Aug 2015 12:08 pm, "Julien Lhermitte" <notifications@github.com>
wrote:

> Two votes is good for me. Submitted. Tested on this branch with same code
> mentioned above.
>
> (math : new axis position: (xa -/+.5*scl + (.5-xa)*scl, then use
> matplotlib functions to convert to data, which take care of non-linear
> plots.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4970#issuecomment-133062305>
> .
>

ordirules commented:
enjoy! :-)

fariza commented:
@ordirules 
I modified your example a little bit to use toolmanager, `ZoomPanBase` was never supposed to be a tool on it's own (if needed we can change that)
```
import matplotlib
matplotlib.use('GTK3AGG')
matplotlib.rcParams['toolbar'] = 'toolmanager'
import matplotlib.pyplot as plt
import numpy as np

fig1 = plt.figure(0)
img = np.random.random((100,100))
ax = fig1.add_subplot(111)
x = np.linspace(10,1000,1000)
y = 1/x**4

#uncomment the one you want to test, and comment the other
#ax.imshow(img)
ax.loglog(x,y)

plt.show()
```

As far as I can test, it works fine!

:+1:

OceanWolf commented:
Ahh, just seen this, a bit busy right now, I will take a look, out of the options, definitely not option 2.  Personally I prefer option 3, but I think option 1 also has merits, I think we can revisit it later though (either RcParam or my personal favourite a toggle tool to set zoom mode.

@fariza Type Agg with Camel case.

ordirules commented:
Thanks @fariza for the test code. The previous code was me desperately trying to find a way to test it with my limited knowledge :-D. Glad I could contribute something, even though exceedingly trivial :-D

@OceanWolf 
One could easily add a flag for going between 1 and 3, maybe something (add in a kargs and pop it into self.centered):
```python
if(self.centered is not None):
    xa,ya=0
else:
    tranP2A = ax.transAxes.inverted().transform
    xa, ya = tranP2A((event.x, event.y))
```
or something of the sort. It might never be used but could be good to add. Is it worth it? I could also do it if it saves you time.

fariza commented:
@ordirules  `ToolZoom` is doing something similar with the `x` and `y` modifiers for `zoom_mode`.
Just as an idea, what about implementing modifiers in the base class?

ordirules commented:
That makes a lot of sense, very good point. Didn't notice that the others inherit from ZoomPanBase. ToolZoom looks redundant with transScale (probably preceded it ;-) ) so it would make sense to clean up the duplicate code and follow the current [transformations philosophy] (http://matplotlib.org/users/transforms_tutorial.html), from my understanding of the tutorial.

The function could be something like "expandshift(ax,x0,y0,x1,y0)" where x,y are the limits in axes coordinates of the new viewbox. I prefer this over "changeview" or something because it emphasizes the relative nature (and potential dangers of looping through).

I think it would also make sense to go deeper and have it part of _AxesBase. The main motivation for this is that it would make life easier for a developer who would want to write their own custom routine without using the GUI. From what I understand (correct me if I'm wrong), but all these inherit from ToolBase which is inherently a GUI feature.

Anyway, what do you think? To sort of summarize:
1. Does such a method already exist to change Axes in _AxesBase?
2. Do you think it's a good idea to add this feature to axes base?

I'll do some reading tonight which may answer my first question (currently busy), but I figured I would ask. Anyway, this is a very good point and I vote for your idea (either case, if it goes in Axes or ZoomPanBase).

ordirules commented:
Hi,

   So I took a look at the _AxesBase class and the only function available seems to be set_xlim etc. I think zooming and panning would be appropriate there, since it is more a feature of the axes and not the GUI. (I could imagine other users using this feature without the GUI)

As a suggestion, I would like to propose adding these features to the _AxesBase class:
zoompanx, zoompany, zoompan
zoompan would work as follows, it would zoom the corresponding axis then pan before setting the limits. (We could also add separate zooming and panning features but I think they're often used together and it might add too many functions)
Afterwards, the update of the zooming and panning GUI tools should be straightforward.

@fariza, @OceanWolf, @tacaswell  this is a slightly more fundamental proposed change. What are your inputs? I can make the changes no problem but need to know if it is commensurate with current matplotlib philosophy.

thanks! :-)

ordirules commented:
Another note, the main reason to add this to the base class is that it requires transformation from axes coordinates to data. The idea would be to allow the user to have a set of routines to change limits in axes coordinates, without worrying about the transformations to data coordinates. I have made the assumption that the only time you want to change limits in axes coordinates is when you pan or zoom, so I've named it zoompan instead of set_axes_lim or something.

OceanWolf commented:
Hmm, before I said I agreed with you, but now not so sure... basically as I see it zoom and pan basically do the same thing, they change the axes limits and they work in a user interactive way... so I think we need to just need a conversion function... perhaps something similar to ``Basemap.__call__`` to convert data coordinates into screen coordinates, but the other way around??

Then again, with different axes types (e.g. 3d) perhaps it becomes easier to do it in the axes itself as it allows the axes to specify its own implementation, rather than making the tool overly complex, catering for all of the different axes types, @WeatherGod you know more about this...

ordirules commented:
Sounds good. I definitely agree with your disagreement with the zoompan idea. :-) 
Just to point out a small issue, it seems that drag_pan in _AxesBase does zooming, and it computes everything in data coordinates as well so it might be something to re-evaluate in the future (or maybe not, because it just works well).

Anyway, I want to bring the focus back (sorry for straying away). This was meant to be a patch to fix log-log plots where the zooming currently breaks down and it is currently fixed. How about we make a decision on how to patch it which I believe is one of two options:
1. Leave as is. Fixed point in zoom is cursor point as well. If we really decide to add a zoom center maybe it should be opened as a feature pull request, not patch pull request?
2. Add a base method in ZoomPanBase that ToolZoom which inherits it can also call (the _release method is a little messy, doing the same thing done in scroll_zoom). The overhead makes me more inclined to have it a feature pull request.

I would say to just leave as is for now and open a new pull request that suggests more fundamental changes. If you agree I think we should just commit. What do you think?

thanks again for the feedback!

ordirules commented:
Didn't work for polar coordinate transforms, so changed the way the transformations are done. Also updated ToolZoom to use the transforms available rather than compute them again. There is still an issue w/ the polar coordinate transformation. (I did not realize that commiting to my github would show up here)

Here are the assumptions I make:
* the transAxes transformation is a transformation from the canvas to the normalized bounding box of the coordinates, and not really the axes (true in projections/polar.py as well, see _set_lim_and_transforms)
* there will always be a transformation transAxes and transData, but there is not always a transformation from the normalized bounding box to the data (transLimits, does not exist on polar coordinates)

As of now, the zooming in polar coordinates isn't perfect with the scroll wheel. Sometimes zooming out with the scroll wheel zooms in. Going through all this code is a learning exercise for me. I apologize for all the writing for such a small patch.

Testing code:
```python
import matplotlib
matplotlib.use('GTK3Agg')
matplotlib.rcParams['toolbar'] = 'toolmanager'
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import numpy as np

plt.ion()

fig1 = plt.figure(0)
img = np.random.random((100,100))
ax = fig1.add_subplot(111)

fig2 = plt.figure(1)
ax2 = fig2.add_subplot(111)

fig3 = plt.figure(2)
ax3 = fig3.add_subplot(111,polar="True")

#fig4 = plt.figure(3)
#ax4 = fig4.add_subplot(111,projection="3d")

x = np.linspace(10,1000,1000)
y = 1/x**4

N = 1000
x = np.linspace(-N/2,N/2,N)
y = np.linspace(-N/2,N/2,N)
X,Y = np.meshgrid(x,y)
w = np.where(np.sqrt(X**2 + Y**2) < 10**2)
img = np.zeros((N,N))
img[w] = 1

#uncomment the one you want to test, and comment the other
ax.imshow(img)

ax2.loglog(x,y)

ax3.plot(x,y)
```

OceanWolf commented:
@QuLogic did some work on polar zooming in #4699.  I don't have time to compare now, but perhaps you should take a look and compare notes.

QuLogic commented:
Maybe you are referring to #4795 actually (which having just been merged is probably causing conflicts here)? Probably `ZoomPanBase` should use the same API added there. I never tried to update it because I could never get it to work. But of course now I realize I should have used the new toolbar to trigger it...

ordirules commented:
Thanks @QuLogic, I think changing the view from a bounding box is
exactly what is needed!!!! :-) Thanks @OceanWolf for the reference.

I've tried my best to reconcile my code with @QuLogic, and in the
meantime made some small modifications. It works for an image, loglog plot and a simple polar coordinate plot. I think the basic idea is that zooming and panning should be done in the bounding box coordinates, and not the data coordinates. Anything else (saving limits, getting limits, like @QuLogic 's getview and setview can only be done in data coordinates).

Here are the major two changes performed:
* Scroll wheel zooming now works in log-log plots, by simply using @QuLogic's method _set_view_from_bbox.
* I removed the check for the log scaling in the _set_view_from_bbox method and replaced it with the axes transformations in transData. The rectangle selection zooming would be affected by any bugs in this change. I assume that a zoom out with this mode just zooms out by a scale which is the ratio of the widths of the rectangle and the view box, and centers the center of the view box. Honestly, I didn't have time to try to check out to see what the mathematical manipulation in the old code exactly meant but it seems about right?

That's about it.

Here is a quick description of the math for the zooming coordinate calculations which are later 
input into @QuLogic's method below. I have not put this in the comments (should I?)
* xcen,xcen (bounding box cen)
* xzc, yzc, center of the selection box (zoom center)
* xd, yd : the data under cursor
    When we select a bounding box, we will basically    
    expand/contract by the amount and re-center:
    xdnew = scl*(xd-xzc) + xcen
    but we don't want xdnew to move so set xdnew = xd
    and solve to get:
    xzc = xd*(scl-1) + xcen
    same for ycen

Finally, the test code I have used has not changed from the previous code.
@QuLogic , could you test it rigorously for polar coordinates?

@fariza, so what do you think?

ordirules commented:
I should also point out on line 3339  (x0, y0, x1, y1 = original_view) from axes/_base.py I had to flip y0 and x1. @QuLogic did zoom work for you with this reversal?

QuLogic commented:
> I should also point out on line 3339 (x0, y0, x1, y1 = original_view) from axes/_base.py I had to flip y0 and x1.

Oh, looking at it, you're probably right; these have been swapped.

QuLogic commented on lib/matplotlib/backend_tools.py:
```diff
@@ -622,28 +621,38 @@ def scroll_zoom(self, event):
+            scl = 1
+
+        ax = event.inaxes
+        xmin, xmax, ymin, ymax = ax._get_view()
```
The point of _get_view/_set_view is to hide the meaning of the "view" from outside of the Axes, because it is not necessarily Cartesian. The following calculation based on it should be placed in the Axes.

ordirules commented:
Hmm, yes I see your point, sorry, I didn't read your pull request completely carefully :-).

My problem here is that I would like to keep zoom working in display space because that's where the zoom is happening. One way is to compute the new bounding box from the old, and only then do you convert to data coordinates to change the limits, just as you did for ToolZoom. However,  to begin, I need to know what the initial bounding box is in display coordinates.
Perhaps if axes has a method to change the view from a new bounding box it should have a method to return the current bounding box?

Another option is to assume that the Axes coordinates are the normalized bounding box coordinates, so 0,0 and 1,1 are the points that delimit the bounding box. Then I can replace _get_view by:
```python
        tranA2C = ax.transAxes.transform
        xmin, ymin = tranA2C((0, 0))
        xmax, ymax = tranA2C((1, 1))
```
However, this does not work in polar coordinates. What is transAxes interpreted as in polar coordinates?

I have searched a little and have not found another alternative. If anyone knows one, please let me know.

QuLogic commented:
Polar axes are a little weird; they implement some strange radial zooming thing. When all the required PRs are merged, #4699 + [this branch](https://github.com/QuLogic/matplotlib/tree/polar-limits-redux) should convert it to a more normal Cartesian zoom.

I'm not sure using `transAxes` would work; I originally planned to use it for Cartesian zoom on `PolarAxes`, but it appeared to also move the title and other elements which seemed unwanted.

You may have to add another method like `_set_scaled_view` or change `_set_view_from_bbox` so that it does the scaling for you in some way. Depending on what other people say, you may not have to worry about `PolarAxes` until the other PRs are merged.

OceanWolf commented:
Ughh, brain hurting, just taken a look at the axes code, and it looks unneedingly verbose.  It looks ripe for refactoring to make the logic as clean as MEP22 and MEP27, i.e. making it easy for people to write custom Axes classes and have them seamlessly integrate into MPL.  However this will probably become part of a much larger MEP involving the Artists in general...

ordirules commented:
@QuLogic thanks for the information. I'm not sure what to do then. I need to find a way to get the display coordinates for it to work. I'll take a look at your branch when I can to understand. There must be a way to get the bounding box information without having to write extra code as well. I'll take a look later. I definitely don't want anything I add to conflict with your wonderful work.

@OceanWolf  ok I will revert change from axes (since it's not necessary), just keep change that fixes zooming (i think I got too carried away). I agree it should be eventually fixed either way, looks like redundant code, but not something for this patch. Thanks for the feedback.
Busy now, but probably later tonight...

ordirules commented:
I updated zooming to be done in the safest way (from what I can find) possible. It grabs the xlim and ylim and uses the transData routine to convert to display coordinates. I agree with @QuLogic that xlim and ylim should eventually be removed from the tools so that they can be overridden by more complex axes classes.

I made original_view in _set_view_from_bbox an optional argument (it grabs the current view before changing if not). Anyway, if it's okay let me know. If not, then maybe this was just not meant to be :-).

ordirules commented:
Hi, if I wanted to help contribute to this library, what would be a first step? I would have messaged @fariza and not posted here but I cannot find any means of sending messages.

thanks!

OceanWolf commented:
@ordirules What did you have in mind?  As far as I see you have already taken the first step by creating this PR :).

We have lots of issues and PRs open.  For big issues we have MEPs (Matplotlib Enhancement Proposals), generally issues that we want to subdivide into smaller PRs and/or an issue that really needs a thought-through strategy.  These I have found quite fun, working on MEP22,27s (refactoring the backends, which me and fariza will continue coding on once we have the next release out), and I have just started an MEP (last week) to look at refactoring the Axes, which still lies in the conceptual stage...

fariza commented:
@ordirules please rebase this PR

ordirules commented:
@OceanWolf thanks I'll take a look. If I have questions, is there anywhere I can post messages?

@fariza sorry for the late reply. my branch was a mess because I am still new to git. I wanted to do a little more [reading](https://git-scm.com/book/en/v2) before rebasing. I had two branches, the original one which was a continuation of my master branch and then a second branch I named newzoom. I had then merged the two together... Given the original master branch resulted in code that I had to remove with @QuLogic's change, I just removed it completely from the history. I forgot to squash my two commits together but I'll try to remember that for next time. 

Anyway, that's a long paragraph just to say I've rebased it with the updated changes and it's ready to merge. Let me know if there's anything else that needs to be fixed.

OceanWolf commented:
I usually ask questions in the issue/PR, but we also have the developer mailing list https://mail.python.org/mailman/listinfo/matplotlib-devel

ordirules commented:
Thanks. I guess before this PR is merged, I should summarize just so it's documented somewhere, it does two things (minor, but more documentation the better):
1. Fix ZoomPanBase to work with log-log plots
2. Very minor fix of @QuLogic's nice addition in two steps:
     A. I allow original_view to be an optional parameter. If not set, use his _get_view command (still within the axes class) to get the current view internally.
     B. Swapped y0 and x1 because it was a typo. (bbox is (xmin,ymin, xmax, ymax) but the views are (xmin,xmax,ymin,ymax) in this current implementation)
      ```python x0, y0, x1, y1 = original_view```
      to
     ```python x0, y1, y0, y1 = original_view```

fariza commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3349,6 +3347,11 @@ def _set_view_from_bbox(self, bbox, original_view, direction='in',
+            x0, x1, y0, y1 = original_view
+        else:
+            x0, x1, y0, y1 = Xmin, Xmax, Ymin, Ymax
+
```
I understand but I think this little change deserves another PR. Unless I'm mistaken this is not related to the `Zoom` problem

ordirules replied:
ok i will add it as another PR

fariza commented on lib/matplotlib/backend_tools.py:
```diff
@@ -626,28 +625,44 @@ def scroll_zoom(self, event):
+        # get the bounding box
+        transD2C = ax.transData.transform
+        xmin, xmax = ax.get_xlim()
+        ymin, ymax = ax.get_ylim()
```
Why don't you use `ax._get_view()` here?
The idea is for this tool to be compatible with all kind of axes that change the view limits.

ordirules replied:
I had a short discussion with @QuLogic about this before. The reason for him adding _get_view was to keep the use of the definition of the view outside the axes. So if one created a new class, like polar axes, one could redefine how the _view methods worked and even what they returned without worrying of dependencies. In polar axes, x,y make less sense. This is why I did not use _get_view.

The best compromise I could find is to assume there exists a two way mapping between canvas (x,y) <--> data(x,y) (the transD2C function). I've tested this for polar coordinates as well and it works, but this might not necessarily be true in the future. (The way I see the zoom currently working I'm guessing it treats one axis as r and the other is ignored)

Probably the best option after that would be to have a zoom option in the new _set_view_from_bbox method, as QuLogic suggested.
However, before doing so, I think the definitions of data, axes and canvas coordinates might need to be discussed revisited. Currently, I don't think the axes coordinates are well defined for systems like polar coordinates, documentation [here](http://matplotlib.org/users/transforms_tutorial.html).

Should we just fix it like this for now and make a bigger change later? I feel like modifying _set_view_from_bbox is a bigger change then just rewriting the currently used code (which uses get_xlim() etc).

fariza commented on lib/matplotlib/backend_tools.py:
```diff
@@ -626,28 +625,44 @@ def scroll_zoom(self, event):
+        bbox = [xzc - xwidth/2./scl, yzc - ywidth/2./scl, xzc +
+                xwidth/2./scl, yzc + ywidth/2./scl]
+
+        ax._set_view_from_bbox(bbox, direction=direction)
```
Maybe I'm asking for too much, but... what about detecting a continuous wheel scroll, and saving the final view in the `ToolViewsPositions` (history) to be able to use it with the `back` and `forward` tools

ordirules replied:
I did not know of this but indeed, it would be a nice idea to save the scroll in the history.
And yes, I definitely agree that some mechanism should be done to figure out when the scroll has started and finished, and not save every scroll in the process! :-)
I'll take a look, unfortunately some time this week. I am busier than expected. :-(

OceanWolf replied:
Remember that we don't save scroll information, just the x and y limits of the axes, that enables both pan and zoom to interact similarly with the core logic.

ordirules commented:
okay I made the modification but before I make the change here we need to accept this PR:
https://github.com/matplotlib/matplotlib/pull/5108

I need this PR to be accepted. It corrects a typo and adds in a default in the event an original_view is not supplied.

OceanWolf commented:
@ordirules I see merge commits above... I think you have made a mistake with git... to rebase one does:
1. Update your local copy of master (see http://matplotlib.org/devel/gitwash/development_workflow.html#staying-up-to-date-with-changes-in-the-central-repository)
2. Checkout your branch and rebase with:
    ```
    git rebase master
    ```

The rebase command "replays" your commits over master so that your branch acts as a transparently over the changes in master requiring no extra commits.  If merge errors occur you have to remember the "replay" rule, i.e. fix only the conflict(s) that arise with the specific commit it currently replays, as it probably has more commits to replay, which might touch the same lines, so you need to leave those changes for later ;).  A slightly different workflow for fixing conflict(s) during rebase, but it gives you instructions, i.e. git add, git rebase --continue; rather than git add, git commit (git status should give you more info).

ordirules commented:
@OceanWolf I see thanks. I'll look  There is no way to cleanly go back is there? Is it better I just close this one and start a new branch and PR from that?

OceanWolf commented:
@ordirules yes,

Check out this post http://stackoverflow.com/questions/1338728/delete-commits-from-a-branch-in-git

For everything else git rebase serves wonders, editing commits, or squashing multiple commits into one, etcetera...

ordirules commented:
I honestly was a little lost on this. I removed a bunch of commits on my local machine, but when I pushed to origin, it asked me to fetch from origin first, and it retrieved all my previous commits. I would like to remove them, is that possible? If it is, I'll spend more energy digging to find out what I did wrong.
Anyway, I'm learning more and more about git which is fun. I'm starting to better understand rebase as well, how it can be used to delete and squash commits, nice feature.

So is this commit history okay or is there something else to do?
The code works and does the following:
  1. Zooming is fixed for log log plots 
  2. The zooming history is saved for scroll events that have been performed more than .5 seconds apart (that's my threshold, if you prefer a different value let me know).

I'll leave a note by the code as well concerning one small point that bothers me.

Thanks for the help, it's been fun learning how this all works. Maybe when I understand it well enough I can help better contribute.

ordirules commented on lib/matplotlib/backend_tools.py:
```diff
@@ -626,30 +628,56 @@ def scroll_zoom(self, event):
+        ax = event.inaxes
+        # note: This code needs to be eventually replaced by axes independent
+        # code to obtain the bounding box coordinates.
+        xmin, xmax = ax.get_xlim()
```
I am still using get_xlim() and get_ylim() which depends on an orthogonal set of axes here. The main reason is that there is no way to obtain the bounding box.
If I want an axes independent zoom, I either:
1. need to obtain the bounding box coordinates from a function that retrieves it (in the axes perhaps)
2. need to call a function that will do the zooming for me (again probably put in axes)

I believe that without any one of these two possibilities (which I do not see as of now), it is not possible to make a tool that zooms and is independent of the choice of axes.
My personal suggestion is to enforce that the axes coordinates be normalized bounding box coordinates. I don't see why axes coordinates would otherwise be useful for since there is already data coordinates.

There are currently three coordinate transforomations: canvas, axes and data. This would mean that canvas *and* axes would be some universal coordinate system that new coordinate systems should conform to while the data coordinates could be left arbitrary (since we would not longer need "set_xlim" and "set_ylim").

So I am asking, is there discussion on the transformations? Given there is more development on coordinate systems such as polar coordinates, I think it's worth revisitng and defining them more rigorously. This is the main reason why I am stuck on using "get_xlim()" thanks.

fariza commented:
@ordirules I think the most logical thing to do is to add `_get_bbox_from_view` to axes, or something like that.
@QuLogic what do you think? This way, axes is "complete" regarding setting and getting view stuff/

OceanWolf commented:
@ordirules the thing about rebase lies in that it changes the hashes of the commit, and when pushing or whatever it obviously needs to know where to hook commits onto, if it sees a different set of commits it gets confused... it doesn't know what's new, what's old, etcetera.

To resolve this you need to use this instead:
```
git push --force
```

This forces git to acknowledge your local version as the correct version.

ordirules commented:
@fariza - I want to wait for [the small view PR] (https://github.com/matplotlib/matplotlib/pull/5108) since it sort of overlaps. (sorry for the delay)

It might be good to also ignore the bbox completely and have the scroll zooming done in _set_view_from_bbox. The idea of the view as @QuLogic mentioned is to keep its handling inside the axes. The set_view_from_bbox routine in my opinion (correct me if I'm wrong) was written simply to allow the selection box tool to work. With that logic, we could tweak it to allow the zoom tool to work as well, rather than have a routine that returns a bounding box from the view, which as far as we know, could be gibberish depending on the axes.


@OceanWolf thanks. I'm  a dummy, I should have thought of that... I think I am now mastering git, it's not bad once you finally get the hang of it. Will clean up the commits again after the PR mentioned gets resolved.

OceanWolf commented:
I only learnt this a year ago (mainly through the use of search engines, and when that failed, asking here).

fariza commented:
@ordirules I think it makes sense to do it inside `_set_view_from_bbox`

ordirules commented:
modifications done but I am a little lost on git, will try to understand this later. I removed all my commits and put them into one and forced a push onto my github.

Lines in axes/_base.py up to line 3215 are not my changes. I must have accidentally not have merged something properly, will look later.

But this is the idea:
    The zooming is now in the axes class _set_view_from_bbox
    It modifies how the bbox tuple is treated. It it's a 3 tuple instead of 4tuple, treat it as zooming. If not, treat it as literal.

Also, the history is saved.

Here is an example to run this on:
```python
import matplotlib
matplotlib.use('GTK3Agg')
matplotlib.rcParams['toolbar'] = 'toolmanager'
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import numpy as np

plt.ion()

fig1 = plt.figure(0)
img = np.random.random((100,100))
ax = fig1.add_subplot(111)

fig2 = plt.figure(1)
ax2 = fig2.add_subplot(111)

fig3 = plt.figure(2)
ax3 = fig3.add_subplot(111,polar="True")

#fig4 = plt.figure(3)
#ax4 = fig4.add_subplot(111,projection="3d")

x = np.linspace(10,1000,1000)
y = 1/x**4

N = 1000
x = np.linspace(-N/2,N/2,N)
y = np.linspace(-N/2,N/2,N)
X,Y = np.meshgrid(x,y)
w = np.where(np.sqrt(X**2 + Y**2) < 10**2)
img = np.zeros((N,N))
theta = np.linspace(0, 2*np.pi,100)
r = theta*0 + 10
img[w] = 1

#uncomment the one you want to test, and comment the other
ax.imshow(img)

ax2.loglog(x,y)

ax3.plot(theta,r)

#ax4 = plot(x,y)
```

Will need to figure out the code cleanup later, but this is the proposed patch.

ordirules commented:
Commit issue fixed. Here's a summary:
* bbox now can be either a 4-tuple (original interpretation) or a 3-tuple. If a 3-tuple, treat it as a xc,yc,scl triplet, where xc,yc are the zoom center (cursor) coordinates and scl is the scale factor.
* zooming has been moved out of the tool and into the axes, _set_view_from_bbox
* Added history for scrolling
* Added two class variables to only save scrolling history within a certain time threshold.
   ```self.scrollthresh``` and ```self.lastscroll```
   where the first is the time threshold in seconds (default 0.5 seconds) and the second is the timestamp of the last scroll

ordirules commented:
and thanks @OceanWolf you really helped me get used to git :-D

jenshnielsen commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3322,8 +3322,12 @@ def _set_view_from_bbox(self, bbox, direction='in',
 
-        bbox : tuple
-            The selected bounding box limits, in *display* coordinates.
+        bbox : 4-tuple or 2 tuple
```
Should this read 4-tuple or 3-tuple?

ordirules replied:
yes it's 3 tuple, oops, thanks!

jenshnielsen commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3342,15 +3346,48 @@ def _set_view_from_bbox(self, bbox, direction='in',
+        if(len(bbox) == 3):
+            # Zooming code
+            xp, yp, scl = bbox
+            print("in set view, for zooming with scale {}".format(scl))
```
Leftover debug output?

ordirules replied:
I thought I had removed all of them. Removed it. As you can tell I didn't recompile and test after removing them. I confirm this time I recompiled and tested this last commit. Thanks!

jenshnielsen commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3342,15 +3346,48 @@ def _set_view_from_bbox(self, bbox, direction='in',
+
+            # Should not happen
+            if(scl == 0):
+                scl = 1.
```
Add a warning?

ordirules replied:
yes, good idea. I'll just print a warning and return. Since it's harmless, I think there's no need to raise any exceptions. However, in that case, the user will see quite a few print statements if they repeat the call with wrong (or even just void) arguments. I'm still new to python (been 4 months, coming from yorick, matlab like open source lang and C), so I'm not completely familiar with the pythonic way of error and warning reporting. Open to recommendations.

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -487,14 +487,14 @@ def __init__(self, fig, rect,
                 sharex._adjustable = 'datalim'
-                #warnings.warn(
+                # warnings.warn(
                 #    'shared axes: "adjustable" is being changed to "datalim"')
```
I think we should just remove these.

ordirules replied:
I just fixed for it to conform with PEP8 but the code is out of scope of this PR.
Maybe make a new PR (and I should probably remove the whitespace PEP8 fix)? @fariza ?

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3322,8 +3322,12 @@ def _set_view_from_bbox(self, bbox, direction='in',
-        bbox : tuple
-            The selected bounding box limits, in *display* coordinates.
+        bbox : 4-tuple or 3 tuple
+            * If result is a 4 tuple, it is the selected bounding box limits,
```
Should be `bbox`, not `result`?

ordirules replied:
yep thanks for the catch

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3322,8 +3322,12 @@ def _set_view_from_bbox(self, bbox, direction='in',
+        bbox : 4-tuple or 3 tuple
+            * If result is a 4 tuple, it is the selected bounding box limits,
+                in *display* coordinates.
+            * If result is a 3 tuple, it is an (xp, yp, scl) triple, where
```
Same here.

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3322,8 +3322,12 @@ def _set_view_from_bbox(self, bbox, direction='in',
+            * If result is a 4 tuple, it is the selected bounding box limits,
+                in *display* coordinates.
+            * If result is a 3 tuple, it is an (xp, yp, scl) triple, where
+                (xp,yp) are the center of zoom and scl the scale factor to zoom
```
"is the center of zooming"

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3322,8 +3322,12 @@ def _set_view_from_bbox(self, bbox, direction='in',
+                in *display* coordinates.
+            * If result is a 3 tuple, it is an (xp, yp, scl) triple, where
+                (xp,yp) are the center of zoom and scl the scale factor to zoom
+                by
```
Missing period.

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3342,15 +3346,52 @@ def _set_view_from_bbox(self, bbox, direction='in',
+        Xmin, Xmax = self.get_xlim()
+        Ymin, Ymax = self.get_ylim()
+
+        if(len(bbox) == 3):
```
Unnecessary parentheses (plus various cases below.)

ordirules replied:
fixed. thanks.

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3342,15 +3346,52 @@ def _set_view_from_bbox(self, bbox, direction='in',
+            xzc = (xp*(scl - 1) + xcen)/scl
+            yzc = (yp*(scl - 1) + ycen)/scl
+
+            bbox = [xzc - xwidth/2./scl, yzc - ywidth/2./scl, xzc +
```
It makes more sense to break after the comma, not the addition.

ordirules replied:
that's a good idea thanks, shouldn't trust my blind vim linewrap :-D

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3342,15 +3346,52 @@ def _set_view_from_bbox(self, bbox, direction='in',
+                    xwidth/2./scl, yzc + ywidth/2./scl]
+        elif(len(bbox) != 4):
+            # should be len 3 or 4 but nothing else
+            print("Warning in _set_view_from_bbox: bounding box is not a\
```
Should be an actual warning, not a print out.

ordirules replied:
kk, followed syntax of others ```warning.warn()```

QuLogic commented on lib/matplotlib/backend_tools.py:
```diff
@@ -626,30 +628,31 @@ def scroll_zoom(self, event):
+        ax = event.inaxes
+        ax._set_view_from_bbox([event.x, event.y, scl])
+
+        # first check if last scroll was done within
```
These sentences could be combined into one.

ordirules replied:
re-worded into something more terse.

QuLogic commented on lib/matplotlib/backend_tools.py:
```diff
@@ -626,30 +628,31 @@ def scroll_zoom(self, event):
+        # first check if last scroll was done within
+        # the timing threshold. If yes, delete previous view
+        if(time.time()-self.lastscroll < self.scrollthresh):
+            # remove the last view and push the current
```
This comment is redundant.

ordirules replied:
removed. thanks for all the feedback!

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3342,15 +3346,52 @@ def _set_view_from_bbox(self, bbox, direction='in',
+            xzc = (xp*(scl - 1) + xcen)/scl
+            yzc = (yp*(scl - 1) + ycen)/scl
+
+            bbox = [xzc - xwidth/2./scl, yzc - ywidth/2./scl, 
```
There's a space at the end here causing PEP8 failures.

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -3342,15 +3346,52 @@ def _set_view_from_bbox(self, bbox, direction='in',
+        elif len(bbox) != 4:
+            # should be len 3 or 4 but nothing else
+            warnings.warn('Warning in _set_view_from_bbox: bounding box is not a\
+                  tuple of length 3 or\ 4. Ignoring the view change...')
```
Not sure why you have a `\` after 'or'.

ordirules replied:
yeah, an artifact of wrapping the text again. I have to be careful when doing so. thanks!

ordirules commented:
Thanks @QuLogic. Travis still failing, I'll need to check tonight.
Also as a test noticed a div by zero error and overflow so will catch the exception (and any other region where this could occur) just to be cleaner. has to do with log log plotting and zooming out too much

jenshnielsen commented:
I think the last Travis error was just a random fluke. I have restarted

ordirules commented:
Great thanks! I never got to catching and avoiding (by not trying to zoom) the div by zero and overflow exceptions. However, these were not checked for in the original code and it could make sense to have this be a separate PR. Do you think it's worth it?

fariza commented:
Is it easy to reproduce?

ordirules commented:
one of them is easy by just zooming in/out way too much. The other I'm not so sure. But I assume by carefully examining it we could probably find a way to test it.

ordirules commented:
oh and thanks @jenshnielsen for the fix and notice!

fariza commented:
@ordirules in that case, a new PR is perfect

OceanWolf commented:
Can you ping me on that, I feel curious to see how you do that.

ordirules commented:
yeah will do :-)
